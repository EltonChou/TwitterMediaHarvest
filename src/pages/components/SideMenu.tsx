import React from 'react'

import { HamburgerIcon } from '@chakra-ui/icons'
import { Box, Flex, IconButton, Link, useBoolean, useBreakpointValue, VStack } from '@chakra-ui/react'
import useLocaleVariables from '@pages/hooks/useLocaleVariables'
import { i18n } from '@pages/utils'
import { Link as RouterLink } from 'react-router-dom'

type MenuItemProps = {
  name: string
  target: string
  closeMenu: () => void
}

const NavItem = (props: MenuItemProps) => {
  return (
    <Link to={props.target} as={RouterLink} onClick={props.closeMenu} _hover={{ textDecoration: 'none' }}>
      <Box
        p="0.5em 0 0.5em 1.5em"
        _hover={{ bg: 'rgba(255, 255, 255, 0.33)' }}
        style={{ transition: 'background 300ms' }}
      >
        {props.name}
      </Box>
    </Link>
  )
}

type NavProps = {
  closeMenu: () => void
}

const Nav = ({ closeMenu }: NavProps) => {
  return (
    <VStack spacing={6} align="normal">
      <NavItem name={i18n('options_sidemenu_general')} target="/" closeMenu={closeMenu} />
      <NavItem name={i18n('options_sidemenu_features')} target="/features" closeMenu={closeMenu} />
      <NavItem name={i18n('options_sidemenu_integrations')} target="/integrations" closeMenu={closeMenu} />
      {/* <NavItem name="History" target="/history" />
      <NavItem name="Statistics" target="/statistics" /> */}
      <NavItem name={i18n('options_sidemenu_about')} target="/about" closeMenu={closeMenu} />
    </VStack>
  )
}

const SideMenu = () => {
  const [isActive, setActive] = useBoolean(false)
  const menuL = useBreakpointValue({
    base: isActive ? '0px' : '-240px',
    lg: '0px',
  })
  const menuBoarder = useBreakpointValue({
    base: 'unset',
    lg: '1px solid gray',
  })
  const navFontSize = useLocaleVariables({
    base: '1.5rem',
    en: '2rem',
  })

  return (
    <>
      <Box pos="fixed" top={0} zIndex={'overlay'}>
        <IconButton
          aria-label="Side menu"
          size={'lg'}
          variant="ghost"
          bg="transparent"
          color="white"
          _hover={{ bg: 'rgba(255, 255, 255, 0.05)' }}
          _active={{ bg: 'rgba(255, 255, 255, 0.2)' }}
          icon={<HamburgerIcon />}
          onClick={() => setActive.on()}
          pos={'absolute'}
          top={0}
        />
        <Box
          height={'full'}
          width={'full'}
          pos={'fixed'}
          bg={'blackAlpha.600'}
          style={{ transition: 'background 300ms' }}
          hidden={!isActive}
          onClick={() => setActive.off()}
        ></Box>
      </Box>
      <Flex
        fontSize={navFontSize}
        direction="column"
        width="240px"
        bg="brand.bg"
        left={menuL}
        borderRight={menuBoarder}
        top="0"
        position={['fixed', 'fixed', 'fixed', 'relative', 'relative']}
        height={'full'}
        zIndex={'modal'}
        style={{
          transition: 'left 200ms',
        }}
      >
        <Box height="150px" />
        <Nav closeMenu={setActive.off} />
      </Flex>
    </>
  )
}

export default SideMenu
