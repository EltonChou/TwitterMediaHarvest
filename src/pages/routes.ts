export const enum Path {
  General = '/',
  Features = '/features',
  Integrations = '/integrations',
  History = '/history',
  About = '/about',
}
