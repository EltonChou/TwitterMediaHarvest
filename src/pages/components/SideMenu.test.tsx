/**
 * @jest-environment jsdom
 */
import { Path } from '#pages/routes'
import theme from '#pages/themes'
import SideMenu from './SideMenu'
import { ChakraProvider } from '@chakra-ui/react'
import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import React from 'react'
import { HashRouter, Route, Routes } from 'react-router-dom'

jest.mock('@chakra-ui/media-query', () => {
  return {
    __esModule: true,
    ...jest.requireActual('@chakra-ui/media-query'),
    useBreakpointValue: () => 'inherit',
  }
})

describe('unit test for SideMenu component', () => {
  afterAll(() => {
    jest.resetAllMocks()
    jest.restoreAllMocks()
  })

  it('can navigate', async () => {
    const { container, unmount } = render(
      <>
        <ChakraProvider theme={theme}>
          <SideMenu />
        </ChakraProvider>
        <Routes>
          <Route path={Path.General} element={<h1>general</h1>} />
          <Route path={Path.Features} element={<h1>features</h1>} />
          <Route path={Path.Integrations} element={<h1>integrations</h1>} />
          <Route path={Path.History} element={<h1>history</h1>} />
          <Route path={Path.About} element={<h1>about</h1>} />
        </Routes>
      </>,
      { wrapper: HashRouter }
    )

    const user = userEvent.setup()

    expect(container).toMatchSnapshot()

    const testCase = [
      { navItemTestId: 'nav-item-general', text: 'general' },
      { navItemTestId: 'nav-item-features', text: 'features' },
      { navItemTestId: 'nav-item-integrations', text: 'integrations' },
      { navItemTestId: 'nav-item-history', text: 'history' },
      { navItemTestId: 'nav-item-about', text: 'about' },
    ]

    for (const { navItemTestId, text } of testCase) {
      await user.click(screen.getByTestId(navItemTestId))
      expect(screen.getByText(text)).toBeInTheDocument()
    }

    await user.click(screen.getByTestId('side-menu-burger'))
    await user.click(screen.getByTestId('side-menu-dimmed'))

    unmount()
  })
})
