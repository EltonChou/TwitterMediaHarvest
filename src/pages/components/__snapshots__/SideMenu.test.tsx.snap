// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit test for SideMenu component can navigate 1`] = `
<div>
  <div
    class="css-bnqvgd"
  >
    <div
      class="css-ye05mj"
    >
      <button
        aria-label="Side menu"
        class="chakra-button css-1j83tvo"
        data-testid="side-menu-burger"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M 3 5 A 1.0001 1.0001 0 1 0 3 7 L 21 7 A 1.0001 1.0001 0 1 0 21 5 L 3 5 z M 3 11 A 1.0001 1.0001 0 1 0 3 13 L 21 13 A 1.0001 1.0001 0 1 0 21 11 L 3 11 z M 3 17 A 1.0001 1.0001 0 1 0 3 19 L 21 19 A 1.0001 1.0001 0 1 0 21 17 L 3 17 z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="css-6arntv"
      data-testid="side-menu-dimmed"
      hidden=""
      style="transition: background 300ms;"
    />
  </div>
  <div
    class="css-1hnriqd"
    style="transition: left 200ms;"
  >
    <div
      class="css-2y8fku"
    />
    <div
      class="chakra-stack css-mu32hu"
    >
      <a
        class="chakra-link css-4a6x12"
        data-discover="true"
        data-testid="nav-item-general"
        href="#/"
      >
        <div
          class="css-1iv9rl2"
          style="transition: background 300ms;"
        >
          Translated&lt;options:sideMenu_General&gt;
        </div>
      </a>
      <a
        class="chakra-link css-4a6x12"
        data-discover="true"
        data-testid="nav-item-features"
        href="#/features"
      >
        <div
          class="css-1iv9rl2"
          style="transition: background 300ms;"
        >
          Translated&lt;options:sideMenu_Features&gt;
        </div>
      </a>
      <a
        class="chakra-link css-4a6x12"
        data-discover="true"
        data-testid="nav-item-integrations"
        href="#/integrations"
      >
        <div
          class="css-1iv9rl2"
          style="transition: background 300ms;"
        >
          Translated&lt;options:sideMenu_Integrations&gt;
        </div>
      </a>
      <a
        class="chakra-link css-4a6x12"
        data-discover="true"
        data-testid="nav-item-history"
        href="#/history"
      >
        <div
          class="css-1iv9rl2"
          style="transition: background 300ms;"
        >
          Translated&lt;options:sideMenu_History&gt;
        </div>
      </a>
      <a
        class="chakra-link css-4a6x12"
        data-discover="true"
        data-testid="nav-item-about"
        href="#/about"
      >
        <div
          class="css-1iv9rl2"
          style="transition: background 300ms;"
        >
          Translated&lt;options:sideMenu_About&gt;
        </div>
      </a>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
  <h1>
    general
  </h1>
</div>
`;
